### ISSUE TITLE
Installer fails on Librem 13v2 with AttributeError

### ISSUE THREAD

**Issue by michas2** [QubesOS/qubes-issues#3050]
#### Qubes OS version (e.g., `R3.2`):
R4.0-rc1
#### Affected TemplateVMs (e.g., `fedora-23`, if applicable):
Installer / anaconda

---

### Expected behavior:
It should not give an error message. :)
It should be able to report errors.

### Actual behavior:
As soon as the graphical installer tries to ask for the language it pops up an python error: "'str' object has no attribute 'name'" 

It offers to send a bug report. However it is not trivial to set up networking at this stage of the installer.
After manually setting up networking the installer tries to report the bug to bugzilla.redhat.com but fails because it uses a 'Generic' product with is not available there.

### Steps to reproduce the behavior:
Use a Librem 13v2 boot from an USB key containing Qubes R4.0-rc1.

### General notes:
Qubes R3.2 installer works fine on that laptop.
Qubes R4.0-rc1 installer works fine on a different laptop.

---

**Comment by marmarek** [comment]
Do you remember file name and/or line number from the message? It's in
details section, at the end of first section (scrolling is a bit hard
there).

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by michas2** [comment]
```
07:11:04,582 DEBUG anaconda: running handleException
07:11:04,583 CRIT anaconda: Traceback (most recent call last):

  File "/usr/lib64/python3.5/site-packages/pyanaconda/threads.py", line 251, in run
    threading.Thread.run(self, *args, **kwargs)

  File "/usr/lib64/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)

  File "/usr/lib/python3.5/site-packages/blivet/osinstall.py", line 1175, in storage_initialize
    storage.reset()

  File "/usr/lib/python3.5/site-packages/blivet/threads.py", line 45, in run_with_lock
    return m(*args, **kwargs)

  File "/usr/lib/python3.5/site-packages/blivet/blivet.py", line 271, in reset
    self.devicetree.populate(cleanup_only=cleanup_only)

  File "/usr/lib/python3.5/site-packages/blivet/threads.py", line 45, in run_with_lock
    return m(*args, **kwargs)

  File "/usr/lib/python3.5/site-packages/blivet/populator/populator.py", line 451, in populate
    self._populate()

  File "/usr/lib/python3.5/site-packages/blivet/threads.py", line 45, in run_with_lock
    return m(*args, **kwargs)

  File "/usr/lib/python3.5/site-packages/blivet/populator/populator.py", line 518, in _populate
    self.handle_device(dev)

  File "/usr/lib/python3.5/site-packages/blivet/threads.py", line 45, in run_with_lock
    return m(*args, **kwargs)

  File "/usr/lib/python3.5/site-packages/blivet/populator/populator.py", line 318, in handle_device
    self.handle_format(info, device)

  File "/usr/lib/python3.5/site-packages/blivet/threads.py", line 45, in run_with_lock
    return m(*args, **kwargs)

  File "/usr/lib/python3.5/site-packages/blivet/populator/populator.py", line 345, in handle_format
    helper_class(self, info, device).run()

  File "/usr/lib/python3.5/site-packages/blivet/populator/helpers/formatpopulator.py", line 89, in run
    self.device.format = formats.get_format(type_spec, **kwargs)

  File "/usr/lib/python3.5/site-packages/blivet/formats/__init__.py", line 98, in get_format
    fmt = fmt_class(*args, **kwargs)

  File "/usr/lib/python3.5/site-packages/blivet/threads.py", line 45, in run_with_lock
    return m(*args, **kwargs)

  File "/usr/lib/python3.5/site-packages/blivet/formats/fs.py", line 125, in __init__
    self.update_size_info()

  File "/usr/lib/python3.5/site-packages/blivet/threads.py", line 45, in run_with_lock
    return m(*args, **kwargs)

  File "/usr/lib/python3.5/site-packages/blivet/formats/fs.py", line 297, in update_size_info
    result = self._minsize.do_task()

  File "/usr/lib/python3.5/site-packages/blivet/tasks/fsminsize.py", line 131, in do_task
    raise FSError("failed to get block size for %s filesystem on %s" % (self.fs.mount_type, self.fs.device.name))

AttributeError: 'str' object has no attribute 'name'
```
[anaconda-2017-08-27-08:48:15.249830-1340.tar.gz](https://github.com/QubesOS/qubes-issues/files/1254290/anaconda-2017-08-27-08.48.15.249830-1340.tar.gz)

**Comment by kakaroto** [comment]
Yep, I had the exact same issue, it's a known/old/unfixed issue reported to fedora here : https://bugzilla.redhat.com/show_bug.cgi?id=1400840
From what I could see, it's a bug in the python3-blivet package, it tries to print self.fs.name variable instead of self.fs
I entered the debugger and found that "self.fs.device" is the actual string, and in my case, it was "/dev/sda1", and it looks like for some reason it doesn't like that device since it can't get its block size, which is what causes the error to be printed which caused the crash.
In my case, I have two SSDs (2.5" and M.2), and I wanted to install Qubes on one of the two, i'm sure that if I delete partitions or maybe reformat /dev/sda1, the error will disappear, but I have a system installed on that disc already which I do not want to overwrite as I was planning on installing qubes on the other disc, and I really don't feel like opening up the laptop to remove the 2.5" SSD just to be able to install Qubes on the M.2 SSD before I can put back the other drive into the laptop..

**Comment by arvog** [comment]
I can confirm that reformatting and setting default partitioning during install yields no error message on a Librem 13v2 (with one M.2 in my case).

**Comment by marmarek** [comment]
What is "/dev/sda1" in that failed case? LVM? Or some filesystem (which one)? What is partition table there (mbr or gpt)?

**Comment by darcy** [comment]
@marmarek this is new territory for me, but have some details that might help. I was able to ctrl-Tab to a console from the debugger and run a few commands (typed in manually, so forgive any errors):

```
# fdisk -l /dev/sda
Disk /dev/sda: 465.8 GiB [...]
Disklabel type: dos

Device    Boot    Start    End    Sectors    Size  Id Type
/dev/sda1        2048   4102143   4100096     2G 83 Linux  
/dev/sda2  *  4102144   5664767   1562624   763M 83 Linux
/dev/sda3     5666814 976771071 971104258 463.1G  5 Extended
/dev/sda4     5666816 976771071 971104256 463.1G 83 Linux
```

```
# parted /dev/sda print
Model: ATA Samsung SSD 850 (scsi)
Disk /dev/sda: 500GB
Sector size (logical/physical): 512B/512B
Partition Table: msdos
Disk Flags:

Number  Start   End    Size    Type      File system  Flags
 1      1049kB  2100MB  2099MB   primary   ext4
 2      2100MB   2100MB  800GB   primary   ext4         boot               
 3      2901MB   500GB  497GB   extended
 5      2901MB   500GB  497GB   logical
```

In the debugger the following values displayed when  set:

```
(Pdb) self.fs 
*** AttributeError: 'Ext4FS' object has no attribute 'name'

(Pdb) self.fs.device
'/dev/sda1'

(Pdb) self.mount_type
'ext4'

(Pdb) self.fs.name
'ext4'

(Pdb) self._extract_block_size()
[None]

(Pdb) self._current_info
[None]

(Pdb) self._get_resize_info()
'Estimated minimum size of the filesystem: 828165\n' 

```

**Comment by darcy** [comment]
Here are the steps I used to reformat the drive to get up and running:

```
parted /dev/sda mklabel gpt
parted -a opt /dev/sda mkpart primary ext4 0% 100%
mkfs.ext4 -L datapartition /dev/sda1
```

**Comment by kakaroto** [comment]
Pretty much the same fdisk/parted results for me, and yes, the existing OS was encrypted, but I think it's irrelevant, the /dev/sda1 was not mounted and I can't mount it either (bad superblock), if I hexdump it, I see this kind of stuff : 
```
00000480  00 00 00 00 00 00 00 00  2f 76 61 72 2f 6c 69 62  |......../var/lib|
00000490  2f 70 75 72 65 6f 73 2d  6f 65 6d 2f 74 61 72 67  |/pureos-oem/targ|
000004a0  65 74 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |et..............|
```
and :
```
08808000  01 43 44 30 30 31 01 00  20 20 20 20 20 20 20 20  |.CD001..        |
08808010  20 20 20 20 20 20 20 20  20 20 20 20 20 20 20 20  |                |
08808020  20 20 20 20 20 20 20 20  50 75 72 65 4f 53 20 38  |        PureOS 8|
08808030  2e 30 20 4c 69 76 65 20  20 20 20 20 20 20 20 20  |.0 Live         |
```
So, it looks to me like this partition is probably the one used by the OEM installer for the first install, and it's meant to keep the copy of the pureos installer ISO while the user does the initial install after receiving their machine. I think it might also have been corrupted (either on purpose to avoid booting into it, or as a bug from the OEM installer), which is why it's not mountable and why the Qubes installer complains about it.

You can find the latest OEM images here : https://downloads.puri.sm/oem/2017-08-20/
It should allow others to reproduce the issue and confirm/fix the problem.
(note, booting into that iso will automatically delete your partitions and install itself on (potentially all) your hard drive(s), I don't think there is any "press enter to confirm" prompt, so don't run it on a system that has any data you're not ready to lose).

**Comment by ksoona** [comment]
@darcy after following these steps on both HDDs, I no longer get the Python error but now get one that VT-d isn't available on my processor. Furthermone, I cannot enable VT-d because I can no longer get into the coreboot coreinfo BIOS (it shows lines of errors instead of the BIOS menu).

**Comment by arvog** [comment]
@ksoona For what I know VT-d is not yet enabled in the latest Librem custom coreboot. You cannot do anything about that, unless you want to help @kakaroto with the programming work on this. Of course you can still install the Qubes 4.0 rcs (at least for me it worked) and maybe also the upcoming final, but until there is a VT-d enabled coreboot version for your Librem, I do not think you should consider your Qubes 4 installation to be safe.

**Comment by jonkri** [comment]
[Things are happening.](https://puri.sm/posts/qubes4-fully-working-on-librem-laptops/)

Purism will announce on their blog when an update to Coreboot 4.7 is available. This update will make it possible to install the latest version of Qubes (on both TPM and non-TPM laptops). 😃

**Comment by kakaroto** [comment]
@jonkri note that with the latest rc4 qubes installer, I still got that same error. And coreboot 4.7 is irrelevant in this case as it won't fix the bug in the installer.
> From what I could see, it's a bug in the python3-blivet package, it tries to print self.fs.name variable instead of self.fs

**Comment by honzahosek** [comment]
I got similar traceback as @michas2 right after Qubes R4.0-rc4 installer start. Also on Librem 13v2. It did not happen when I removed the hard disk with PureOS installed on it. I think the bug is fixed here: https://github.com/storaged-project/blivet/commit/b4407b21415f6ede2542ef044fdc5d45b5dd5c48#diff-4f946d685eb25b56e6b3e57fcc5c00d8

**Comment by marmarek** [comment]
But the AttributeError is in the code reporting an error. So, when we fix one, we'll get another (not saying it shouldn't be fixed obviously). Can you post the PureOS partition layout so I could replicate the problem?

**Comment by marmarek** [comment]
Eh, I should read more than last 3 comments before writing mine. I see @darcy already provided partition layout and @kakaroto the link to OEM images...

**Comment by kakaroto** [comment]
Yep, that looks like the fix in blivet, it just needs to be integrated into qubes installer now. And yes, that will not prevent the error due to the partition,but it will prevent a crash of the installer. I don't know what will happen, but I expect a popup or something that says "unknown partition type" somewhere, definitely a recoverable error.

**Comment by marmarek** [comment]
That was it. Before applying the fix: https://openqa.qubes-os.org/tests/93
After: https://openqa.qubes-os.org/tests/112

**Comment by kakaroto** [comment]
Awesome, thanks!

**Comment by marmarek** [comment]
BTW if you want to test it yourself, you can either [build iso](https://www.qubes-os.org/doc/qubes-r3-building/), or download iso from the above linked successful openQA test ("Logs & Assets" tab). That's new part of our testing infrastructure and we're still evaluating possibilities. Note that the iso image there is intentionally not signed - use it for testing only.
